### DISCUSSION TITLE
How to bind sequence to a table

### DISCUSSION THREAD

**Discussion by febus** [linq2db/linq2db#3698] [Q&A]
Hello,
Is it possible to use sequences with linq2db?
I get the following code generated for table
```C#
[Table("my_status")]
public partial class MyStatusTableDto
{
  [Column("id", IsPrimaryKey = true)] public long   Id     { get; set; } // numeric(10,0)
  [Column("group", CanBeNull = false)] public string Group { get; set; } = null!; // character varying(50)
  [Column("text"  , CanBeNull = false)] public string Text   { get; set; } = null!; // character varying(50)
  ...
}
```
Now I want to insert a new row into the table. The question is, how does linq2db know, which database sequence I want to use in order to generate a new value for id column? Is there any way to tell linq2db cli tool which sequence I want to use for my primary key columns?
How can  I bind a sequence to a table? Let's say per convention I'd like to call sequences like S_Table , so for this table the sequence should be called s_my_status. 
Can this be done, with a help of interceptor DLL, for example? Or probably you know some other solution?

**Comment by MaceWindu** [comment]
You can use `SequenceName` attribute, e.g. see https://github.com/linq2db/linq2db/blob/master/Tests/Linq/DataProvider/OracleTests.cs#L3672

Note that it is currently supported only for oracle/pgsql/firebird

**Comment by febus** [comment]
Yes, I've seen this Attribute already. The problem is I've got generated code for 200 tables and even more Views, which is actually unusable for writing, since I can't recreate it easily in case Database changes. Is there any way to put SequenceName on public keys columns properties automatically, while generating classes with linq2db cli tool? I searched in sources, but this seems to be inpossible.

**Reply by MaceWindu** [reply]
I see. Yep, we currently doesn't support sequences metadata load (#2766).

I see only one option that could preserve sequence information after context regeneration - you can use fluent mapping to specify sequences on columns.

**Reply by febus** [reply]
Unfortunately there  is nothing in fluent mapping for adding sequence attribute.
I think in its simplest case this should be something like this:
```C#
public partial class MyDataContext
{
    partial void InitDataContext()
    {
        var mb = this.MappingSchema.GetFluentMappingBuilder();
        mb.Entity<MyTableDto>().Property(_ => _.Idx).HasAttribute(new **SequenceNameAttribute???**("s_my_table"));
    }
}
```

HasAttribute() accepts only some derived instance of MappingAttribute linq2db class.
The problem is, there is nothing derived from MappingAttribute to define SequenceName. All kind of things, but nothing for sequences. Or I'm missing something, which is still possible. HasSequence is also not there :)

**Reply by MaceWindu** [reply]
amazing... will fix for next release #3701

**Reply by febus** [reply]
I added a new configuration option myself to linq2db cli tool. I called it "sequence-name-prefix" and it is a part of dataModel json configuration file. I could also add a "sequence-name-suffix" option, to add a suffix to a table name. So, I suppose that a  sequence is always called by convention [prefix][tablename][suffix]
In my opinion you don't need to read sequences from database schema, since they are not connected in no way to a specific table on the database side, so that you alway should relay on some convention, to connect sequences and corresponding tables.
SequenceNameAttribute is now derived from MappingAttribute also, which allows to use it with a fluent mapping as well.

If specified, this options addes SequenceName attribute to a primary column automatically for all tables in data context.
Probaby you can build it the same way in a future release.
Or I can try a pull request, if you have nothing agaisnt it. But I did everything in "master" now :)

My current problem is, I can't load interceptor dll from debugger, if I set LinqToDB.CLI as a startup project.
Exactly the same interceptor dll works perfectly if used from dotnet linq2db cli tool installed globally.

Actually I only need no set Schema to null without my interceptor dll, then I don't need it really anymore.

I copied my interceptor dll to the same directory as my custom built dotnet-linq2db.exe, since linq2db.tools.dll must be the same, otherwise interceptor dll can't be recognized and loaded.
Where should I place my interceptor dll if I want to use it as a command line argument, while debugging  LinqToDB.CLI project in Visual Studio ? Here is what I'm getting now, while trying to load an interceptor dll from the same directory as a dotnet-linq2db.exe:
![image](https://user-images.githubusercontent.com/935986/184902176-e0fcf144-1a73-44b2-a15f-2e6b3f362c4e.png)

P.S. Я прекрасно понимаю по-русски :)
Спасибо за Вашу работу, наша контора использует Linq2Db годами, но в старой версии, еще под .Net Framework. Сейчас тестирую под .NET 6

**Reply by MaceWindu** [reply]
Sorry, didn't  noticed questions about interceptors..

It would be nice if you provide PR for sequence configuration (tracked by #3749) after https://github.com/linq2db/linq2db/pull/3748 merged. Note that:
- sequence configuration should be a part of `ColumnMetadata` class
- attribute generation implemented in `AttributeBasedMetadataBuilder`
- sequence configuration includes two fields: sequence name and optional sequence schema name

Regarding CLI configuration options. I actually not sure yet which options we need so we can go with your prefix/suffix options as starting point. Probably we will need to add:
- table filtering option to skip/select specific tables (we can use same approach as we use for `include-/exclude-` function filters)
- more complex sequence name generator using replace regex instead of plain prefix/suffix (see kinda similar request https://github.com/linq2db/linq2db/issues/3739)


> Actually I only need no set Schema to null without my interceptor dll, then I don't need it really anymore.

I plan to add required option before release

> My current problem is, I can't load interceptor dll from debugger, if I set LinqToDB.CLI as a startup project.
Exactly the same interceptor dll works perfectly if used from dotnet linq2db cli tool installed globally.

not sure what's the problem here, but best option should be a reference to published dll in separate folder with own `deps.json` file. Maybe you mix build results from different projects (interceptor and cli) and overwrite deps.json?

**Comment by Metadorius** [comment]
I faced the same issue today. I have a DB with mapping of something like 30k lines of code, and I can't add `SequenceName` attributes on every regeneration... Ideally I would want to do something like a partial property in a separate file, but alas, C# devs don't even seem to consider such a possibility. Perhaps there's some workaround solution?

**Reply by febus** [reply]
My solution so far was a custom modification of a linq2db cli tool. I've added a new option sequence-name-prefix which does exactly what its name says: creates a SequenceName attribute on primary key properties with the name <prefix>tablename.
I could create a pull request, if @MaceWindu would be interested in a such functionality.
Actually we should also add not only prefix, but also a suffix. In my other even much bigger database (110k lines of generated code) sequences are called table_seq, so I need a suffix as well ;) 

I also tried to filter out  primary keys in my interceptor dll. The plan was to add them manually, since I generate partial classes.
The problem with this approach is that associations couldn't be created, since I've taken all primary keys away. :)

**Reply by Metadorius** [reply]
Some of the generators in my DB unfortunately don't follow such naming scheme, ideally would want some kind of a map where we can specify which generator should be used for which field.

**Reply by MaceWindu** [reply]
Next release will enable sequences configuration using fluent mapping #3748

**Reply by Metadorius** [reply]
@MaceWindu Actually, is it currently possible to specify the sequence in an interceptor?

**Reply by MaceWindu** [reply]
Not yet. Sequences not yet known to scaffold tool. You can track #3749 for it (I expect it to be in next release)
